gbpservice/nfp/orchestrator/config_drivers/heat_driver.py

```diff
@@ -1161,6 +1456,65 @@ def apply_config(self, network_function_details):
 
         return stack_id
 
+    def apply_heat_config(self, nfp_context):
+        service_details = self.get_service_details_from_nfp_context(
+            nfp_context)
+
+        network_function = nfp_context['network_function']
+        service_profile = service_details['service_profile']
+        service_chain_node = service_details['servicechain_node']
+        service_chain_instance = service_details['servicechain_instance']
+        provider = service_details['provider_ptg']
+        consumer = service_details['consumer_ptg']
+        consumer_port = service_details['consumer_port']
+        provider_port = service_details['provider_port']
+        mgmt_ip = service_details['mgmt_ip']
+
+        auth_token = nfp_context['resource_owner_context']['admin_token']
+        provider_tenant_id = nfp_context['tenant_id']
+        heatclient = self._get_heat_client_v1(provider_tenant_id,
+                                              assign_admin=True)
+        if not heatclient:
+            return None
+
+        stack_template, stack_params = self._create_node_config_data(
+            auth_token, provider_tenant_id,
+            service_chain_node, service_chain_instance,
+            provider, provider_port, consumer, consumer_port,
+            network_function, mgmt_ip, service_details)
+
+        if not stack_template and not stack_params:
+            return None
+
+        if not heatclient:
+            return None
+
+        stack_name = ("stack_" + service_chain_instance['name'] +
+                      service_chain_node['name'] +
+                      service_chain_instance['id'][:8] +
+                      service_chain_node['id'][:8] + '-' +
+                      time.strftime("%Y%m%d%H%M%S"))
+        # Heat does not accept space in stack name
+        stack_name = stack_name.replace(" ", "")
+
+        try:
+            stack = heatclient.create(stack_name, stack_template, stack_params)
+        except Exception as err:
+            LOG.error(_LE("Heat stack creation failed for template : "
+                          "%(template)s and stack parameters : %(params)s "
+                          "with Error: %(error)s") %
+                      {'template': stack_template, 'params': stack_params,
+                       'error': err})
+            return None
+
+        stack_id = stack['stack']['id']
+        LOG.info(_LI("Created stack with ID %(stack_id)s and "
+                     "name %(stack_name)s for provider PTG %(provider)s"),
+                 {'stack_id': stack_id, 'stack_name': stack_name,
+                  'provider': provider['id']})
+
+        return stack_id, heatclient
```

Comment:
Why we are returning heatclient to module. Module should not use/access heatclient.